File: backend/app/api/sessions.py
import uuid
from datetime import date
from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy import desc, select
from sqlalchemy.orm import Session

from app.database import get_db_session
from app.models import Exercise, ExerciseSet, WorkoutSession
from app.schemas import SessionDetailOut, SessionExerciseOut, SessionListItemOut, SessionSetOut

router = APIRouter(prefix="/api/sessions", tags=["sessions"])


def _get_db(request: Request):
    yield from get_db_session(request.app.state.session_factory)


@router.get("", response_model=List[SessionListItemOut])
def list_sessions(
    from_date: date = Query(default=None, alias="from"),
    to_date: date = Query(default=None, alias="to"),
    limit: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(_get_db),
) -> List[SessionListItemOut]:
    stmt = select(WorkoutSession)
    if from_date is not None:
        stmt = stmt.where(WorkoutSession.date >= from_date)
    if to_date is not None:
        stmt = stmt.where(WorkoutSession.date <= to_date)
    stmt = stmt.order_by(desc(WorkoutSession.date), desc(WorkoutSession.created_at)).limit(limit)
    rows = db.execute(stmt).scalars().all()
    return [SessionListItemOut.model_validate(row) for row in rows]


@router.get("/{session_id}", response_model=SessionDetailOut)
def get_session_detail(session_id: uuid.UUID, db: Session = Depends(_get_db)) -> SessionDetailOut:
    session_row = db.get(WorkoutSession, session_id)
    if session_row is None:
        raise HTTPException(status_code=404, detail="session_not_found")

    exercise_rows = (
        db.execute(select(Exercise).where(Exercise.session_id == session_row.id).order_by(Exercise.order_index)).scalars().all()
    )
    exercises: List[SessionExerciseOut] = []
    for exercise in exercise_rows:
        set_rows = (
            db.execute(select(ExerciseSet).where(ExerciseSet.exercise_id == exercise.id).order_by(ExerciseSet.set_index))
            .scalars()
            .all()
        )
        exercises.append(
            SessionExerciseOut(
                id=exercise.id,
                raw_name=exercise.raw_name,
                order_index=exercise.order_index,
                sets=[SessionSetOut.model_validate(set_row) for set_row in set_rows],
            )
        )

    return SessionDetailOut(
        id=session_row.id,
        date=session_row.date,
        calories_kcal=session_row.calories_kcal,
        duration_min=session_row.duration_min,
        volume_kg=session_row.volume_kg,
        upload_id=session_row.upload_id,
        exercises=exercises,
    )

File: cloudflare-api/src/handlers/sessions.ts
import { errorResponse, jsonResponse } from "../http";
import { getBodyweightSetting } from "../repositories/settings";
import { loadSessionSets } from "../repositories/sessions";
import {
  applySessionUpdate,
  cloneSessionTree,
  removeSessionWithChildren,
  resetNonSeedSessions,
  type NormalizedSessionPayload,
} from "../services/session-workflow";
import { normalizeSessionPayload } from "../session-mutations";
import type { RequestContext, Env } from "../types";

type SessionSummaryRow = {
  id: string;
  date: string;
  started_at: string | null;
  calories_kcal: number | null;
  duration_min: number | null;
  volume_kg: number | null;
  upload_id: string | null;
};

function defaultCloneDate(): string {
  return new Date().toISOString().slice(0, 10);
}

export async function listSessions(url: URL, env: Env, ctx: RequestContext): Promise<Response> {
  const fromDate = url.searchParams.get("from");
  const toDate = url.searchParams.get("to");
  const limit = Number.parseInt(url.searchParams.get("limit") || "50", 10);
  const boundedLimit = Number.isNaN(limit) ? 50 : Math.max(1, Math.min(200, limit));

  const clauses: string[] = [];
  const binds: Array<string | number> = [];
  if (fromDate) {
    clauses.push("date >= ?");
    binds.push(fromDate);
  }
  if (toDate) {
    clauses.push("date <= ?");
    binds.push(toDate);
  }

  const whereSql = clauses.length > 0 ? `WHERE ${clauses.join(" AND ")}` : "";
  const sql = `
    SELECT id, date, started_at, calories_kcal, duration_min, volume_kg, upload_id
    FROM sessions
    ${whereSql}
    ORDER BY date DESC, created_at DESC
    LIMIT ?
  `;
  binds.push(boundedLimit);

  const rows = await env.DB.prepare(sql).bind(...binds).all<SessionSummaryRow>();
  return jsonResponse(rows.results, ctx.requestId);
}

export async function getSessionDetail(sessionId: string, env: Env, ctx: RequestContext): Promise<Response> {
  const sessionRow = await env.DB
    .prepare("SELECT id, date, started_at, calories_kcal, duration_min, volume_kg, upload_id FROM sessions WHERE id = ?1")
    .bind(sessionId)
    .first<SessionSummaryRow>();

  if (!sessionRow) {
    return errorResponse(404, "session_not_found", ctx.requestId);
  }

  const exerciseRows = await env.DB
    .prepare("SELECT id, raw_name, order_index FROM exercises WHERE session_id = ?1 ORDER BY order_index ASC")
    .bind(sessionId)
    .all<{ id: string; raw_name: string; order_index: number }>();

  const exerciseIds = exerciseRows.results.map((exercise) => exercise.id);
  const setsByExerciseId = await loadSessionSets(env, exerciseIds);

  const exercises = exerciseRows.results.map((exercise) => ({
    id: exercise.id,
    raw_name: exercise.raw_name,
    order_index: exercise.order_index,
    sets: setsByExerciseId.get(exercise.id) ?? [],
  }));

  return jsonResponse({ ...sessionRow, exercises }, ctx.requestId);
}

export async function updateSession(
  req: Request,
  sessionId: string,
  env: Env,
  ctx: RequestContext
): Promise<Response> {
  const existing = await env.DB.prepare("SELECT id FROM sessions WHERE id = ?1").bind(sessionId).first<{ id: string }>();
  if (!existing) {
    return errorResponse(404, "session_not_found", ctx.requestId);
  }

  let payloadRaw: unknown;
  try {
    payloadRaw = await req.json();
  } catch {
    return errorResponse(400, "invalid_json", ctx.requestId);
  }

  let payload;
  try {
    payload = normalizeSessionPayload(payloadRaw as never, { bodyweight_kg: await getBodyweightSetting(env) });
  } catch (error) {
    return errorResponse(400, error instanceof Error ? error.message : "invalid_payload", ctx.requestId);
  }

  try {
    await applySessionUpdate(env, sessionId, payload as NormalizedSessionPayload);
  } catch (error) {
    console.error(
      JSON.stringify({
        level: "error",
        message: "session_update_failed",
        request_id: ctx.requestId,
        session_id: sessionId,
        error: error instanceof Error ? error.message : String(error),
      })
    );
    return errorResponse(400, "invalid_payload", ctx.requestId);
  }

  return getSessionDetail(sessionId, env, ctx);
}

export async function deleteSession(sessionId: string, env: Env, ctx: RequestContext): Promise<Response> {
  const existing = await env.DB.prepare("SELECT id FROM sessions WHERE id = ?1").bind(sessionId).first<{ id: string }>();
  if (!existing) {
    return errorResponse(404, "session_not_found", ctx.requestId);
  }

  await removeSessionWithChildren(env, sessionId);
  return jsonResponse({ ok: true, deleted_session_id: sessionId }, ctx.requestId);
}

export async function resetSessions(env: Env, ctx: RequestContext): Promise<Response> {
  const result = await resetNonSeedSessions(env);
  return jsonResponse(
    { ok: true, deleted_sessions: result.deletedSessions, deleted_uploads: result.deletedUploads },
    ctx.requestId
  );
}

export async function cloneSession(
  req: Request,
  sessionId: string,
  env: Env,
  ctx: RequestContext
): Promise<Response> {
  const source = await env.DB
    .prepare("SELECT id, date, started_at, calories_kcal, duration_min, volume_kg FROM sessions WHERE id = ?1")
    .bind(sessionId)
    .first<{
      id: string;
      date: string;
      started_at: string | null;
      calories_kcal: number | null;
      duration_min: number | null;
      volume_kg: number | null;
    }>();
  if (!source) {
    return errorResponse(404, "session_not_found", ctx.requestId);
  }

  let requestedDate = "";
  try {
    const body = (await req.json()) as { date?: string };
    requestedDate = String(body?.date || "").trim();
  } catch {
    requestedDate = "";
  }
  const targetDate = /^\d{4}-\d{2}-\d{2}$/.test(requestedDate) ? requestedDate : defaultCloneDate();
  const newSessionId = await cloneSessionTree(env, source, targetDate);

  return getSessionDetail(newSessionId, env, ctx);
}

